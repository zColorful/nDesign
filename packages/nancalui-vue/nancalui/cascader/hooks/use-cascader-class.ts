/**
 * 定义组件class
 */
import { computed, ComputedRef, Ref, toRefs } from 'vue';
import { CascaderProps, CascaderulProps } from '../src/cascader-types';
import { useNamespace } from '../../shared/hooks/use-namespace';
import { useSize } from '../../shared/_utils/use-size';
const ns = useNamespace('cascader');

// 根节点class
export const useRootClassName = (props: CascaderProps, menuShow: Ref<boolean>): ComputedRef => {
  return computed(() => ({
    [`${ns.b()} ${ns.e('dropdown')} ${ns.em('dropdown', 'animation')}`]: true,
    [ns.em('dropdown', 'open')]: menuShow.value,
    [ns.e('disbaled')]: props.disabled,
  }));
};

// 弹出层列 class
export const useUlClassName = (props: any): ComputedRef => {
  return computed(() => ({
    [ns.e('ul')]: true,
    [ns.e('drop-no-data')]: !props?.displayColumns?.length && !props?.options?.length,
  }));
};

// 为弹出层打开添加全局class
export const dropdownOpenClass = (status: boolean): string => {
  return status ? `${ns.e('drop-menu-wrapper')}` : '';
};
