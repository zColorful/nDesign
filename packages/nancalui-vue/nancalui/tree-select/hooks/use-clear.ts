import { computed } from 'vue';
import { TreeSelectProps } from '../src/tree-select-types';

export default function useClear(props: TreeSelectProps): unknown {
  const isClearable = computed<boolean>(() => {
    return !props.disabled && props.allowClear;
  });

  return {
    isClearable,
  };
}
