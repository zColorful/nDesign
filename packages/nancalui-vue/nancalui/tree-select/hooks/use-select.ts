import { ref } from 'vue';
import { TreeSelectProps, TreeItem } from '../src/tree-select-types';

export default function useSelect(props: TreeSelectProps, ctx: SetupContext): unknown {
  let { label, value, children } = props.prop;
  label = label ? label : 'label';
  value = value ? value : 'value';
  children = children ? children : 'children';
  const inputValue = ref<TreeItem | Array<TreeItem>>([]); // 选中的label值
  const selectedValue = ref<unknown | Array<unknown>>([]); // 选中的value值
  const selectedCache = new Set();
  const selectedValueCache = new Set();
  const useCache = (item: TreeItem) => {
    // item.checked === true ? selectedCache.add(item.label) : selectedCache.has(item.label) && selectedCache.delete(item.label);
    if (item.checked === true) {
      selectedCache.add(item);
      selectedValueCache.add(item[value]);
    } else {
      selectedCache.has(item) && selectedCache.delete(item);
      selectedValueCache.has(item[value]) && selectedValueCache.delete(item[value]);
    }
  };
  const searchUp = (item: TreeItem) => {
    if (!item.parent) {
      return;
    }
    let state = '';
    const checkedArr = item.parent.children.filter((el) => el.checked === true);
    switch (checkedArr.length) {
      case 0:
        state = 'none';
        break;
      case item.parent.children.length:
        state = 'checked';
        break;
      default:
        state = 'halfchecked';
        break;
    }

    if (state === 'checked') {
      item.parent.checked = true;
      item.parent.halfchecked = false;
    } else if (state === 'halfchecked') {
      item.parent.halfchecked = true;
      item.parent.checked = false;
    } else {
      item.parent.checked = false;
      item.parent.halfchecked = false;
    }

    useCache(item.parent);
    searchUp(item.parent);
  };
  const clearSelect = () => {
    if (!props.multiple) {
      selectedValue.value = undefined;
      inputValue.value = '';
    } else {
      selectedValue.value = [];
      inputValue.value = [];
    }
    selectedCache.clear();
    selectedValueCache.clear();
  };
  const searchDown = (item: TreeItem) => {
    if (!item.children) {
      return;
    }
    item.children.forEach((el) => {
      el.checked = item.checked;
      useCache(el);
      searchDown(el);
    });
  };

  const selectValue = (item: TreeItem, bloodSearch: boolean) => {
    // bloodSearch 是否在初始化选中节点
    if (!props.multiple) {
      selectedValue.value = item[value];
      inputValue.value = item[label];
    } else {
      if (!bloodSearch) {
        item.checked = !item.checked;
      } else {
        if (!item.checked) {
          item.checked = !item.checked;
        }
      }
      if (item.halfchecked) {
        item.halfchecked = false;
      }
      useCache(item);
      searchUp(item);
      searchDown(item);

      // if (selectedCache.size > 0) {
      inputValue.value = [...selectedCache] as string[];
      // }
      // if (selectedValueCache.size > 0) {
      selectedValue.value = [...selectedValueCache];
      // }
    }
    if (!bloodSearch) {
      ctx.emit('update:modelValue', selectedValue.value);
      ctx.emit('valueChange', { data: selectedValue.value, node: item });
    }
  };
  const clearSelected = (data: Array<TreeItem>) => {
    for (const item of data) {
      item.checked = false;
      item.selected = false;
      if (item[children]?.length) {
        clearSelected(item[children]);
      }
    }
  };
  const handleClearAll = (e: MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    clearSelect();
    clearSelected(props.treeData);
    ctx.emit('valueChange', { data: '' });
    ctx.emit('update:modelValue', '');
  };
  const handleClearItem = (e: MouseEvent, item?: string) => {
    e.preventDefault();
    e.stopPropagation();
    if (props.multiple) {
      const index = inputValue.value?.findIndex((row: TreeItem) => row[value] === item[value]);
      inputValue.value.splice(index, 1);
      selectedValue.value.splice(selectedValue.value.indexOf(item[value]), 1);
    } else {
      inputValue.value = '';
      selectedValue.value = '';
    }
    ctx.emit('valueChange', { data: selectedValue.value, node: item });
    ctx.emit('update:modelValue', selectedValue.value);
  };

  return {
    inputValue,
    selectValue,
    clearSelect,
    selectedValue,
    handleClearAll,
    handleClearItem,
  };
}
