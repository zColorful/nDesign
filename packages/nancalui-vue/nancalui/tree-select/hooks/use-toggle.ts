import { ref } from 'vue';
import { TreeSelectProps, TreeItem } from '../src/tree-select-types';

export default function useToggle(props: TreeSelectProps) {
  const visible = ref<boolean>(false);

  const selectToggle = (val: boolean) => {
    if (props.disabled) {
      return;
    }
    if (val) {
      visible.value = val;
    } else {
      visible.value = !visible.value;
    }
  };

  const treeToggle = (e: MouseEvent, item: TreeItem) => {
    e.preventDefault();
    e.stopPropagation();
    item.expanded = !item.expanded;
  };

  return {
    visible,
    selectToggle,
    treeToggle,
  };
}
