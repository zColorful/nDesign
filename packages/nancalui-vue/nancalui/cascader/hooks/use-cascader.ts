/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable no-use-before-define */
import { ref, SetupContext, toRef, reactive, provide, watch, onMounted, toRefs, computed } from 'vue';
import { cloneDeep, debounce, isFunction, isNull, isUndefined } from 'lodash';
import type { CascaderProps } from '../src/cascader-types';
import { popupHandles } from './use-cascader-popup';
import { useRootStyle } from './use-cascader-style';
import { useRootClassName } from './use-cascader-class';
import { useFormItem } from '../../shared/_utils/use-form-item';
import { CascaderOption, CascaderOptionInfo } from '../src/interface';
import { BaseType, UnionType } from '../../shared/_utils/types';
import {
  getCheckedStatus,
  getLeafOptionInfos,
  getLeafOptionKeys,
  getOptionInfos,
  getOptionLabel,
  getValidValues,
  getValueKey,
} from '../src/utils';
import { useSelectedPath } from './use-selected-path';
import { cascaderInjectionKey } from '../src/context';
import { getKeyDownHandler, KEYBOARD_KEY } from '../../shared/_utils/keyboard';

export const useCascader = (props: CascaderProps, ctx: SetupContext) => {
  const origin = ref<HTMLElement>();
  const overlayRef = ref<HTMLElement>();
  const multiple = toRef(props, 'multiple');
  const inputValue = ref('');
  const rootStyle = useRootStyle(props);
  const showClearable = ref(false);
  const position = ref(['bottom-start', 'top-start']);
  const { emit, slots } = ctx;
  const { disabled, modelValue, options, trigger, loadMore, formatLabel, checkStrictly, valueKey, expandChild } = toRefs(props);

  const _value = ref();
  const _inputValue = ref();

  const { mergedDisabled } = useFormItem({ disabled });
  const computedInputValue = computed(() => props.inputValue ?? _inputValue.value);
  const computedPopupVisible = computed(() => props.popupVisible ?? menuShow.value);
  // popup弹出层
  const { menuShow, menuOpenClass, openPopup } = popupHandles(props, overlayRef, origin);
  // 配置class
  const rootClasses = useRootClassName(props, computedPopupVisible as any);

  watch(modelValue, (value) => {
    if (isUndefined(value) || isNull(value)) {
      _value.value = props.multiple ? [] : undefined;
    }
  });

  const optionInfos = ref<CascaderOptionInfo[]>([]);
  const totalLevel = ref(1);

  const optionMap = reactive(new Map<string, CascaderOptionInfo>());
  const leafOptionMap = reactive(new Map<string, CascaderOptionInfo>());
  const leafOptionValueMap = reactive(new Map<BaseType, string>());
  const leafOptionSet = reactive(new Set<CascaderOptionInfo>());

  const lazyLoadOptions = reactive<Record<string, CascaderOption[]>>({});

  const addLazyLoadOptions = (children: CascaderOption[], key: string) => {
    lazyLoadOptions[key] = children;
  };

  const DEFAULT_FIELD_NAMES = {
    value: 'value',
    label: 'label',
    disabled: 'disabled',
    children: 'children',
    render: 'render',
    isLeaf: 'isLeaf',
  };

  const mergedFieldNames = computed(() => ({
    ...DEFAULT_FIELD_NAMES,
    ...props.fieldNames,
  }));

  //  选项变化直接重新生成map信息
  watch(
    [options, lazyLoadOptions, mergedFieldNames],
    ([_options, _lazyLoadOptions, _fieldNames]) => {
      optionMap.clear();
      leafOptionMap.clear();
      leafOptionValueMap.clear();
      leafOptionSet.clear();

      optionInfos.value = getOptionInfos(_options ?? [], {
        enabledLazyLoad: Boolean(props.loadMore),
        lazyLoadOptions,
        optionMap,
        leafOptionSet,
        leafOptionMap,
        leafOptionValueMap,
        totalLevel,
        checkStrictly,
        valueKey,
        fieldNames: _fieldNames,
      });
    },
    {
      immediate: true,
      deep: true,
    }
  );

  const computedValueMap = computed(() => {
    const values = getValidValues(props.modelValue ?? _value.value, {
      multiple: props.multiple,
      pathMode: props.pathMode,
    });
    return new Map(
      values.map((value) => [
        getValueKey(value, {
          valueKey: props.valueKey,
          leafOptionValueMap,
        }),
        value,
      ])
    );
  });

  const getFilteredStatus = (label: string) => {
    return label?.toLocaleLowerCase().includes(computedInputValue.value?.toLocaleLowerCase());
  };

  const filteredLeafOptions = computed(() => {
    let _a;
    (_a = overlayRef.value) == null ? void 0 : _a.updatePosition();
    const opts = props.checkStrictly ? Array.from(optionMap.values()) : Array.from(leafOptionSet);

    return opts.filter((item) => {
      if (isFunction(props.filterOption)) {
        let _query = '';
        _query = computedInputValue.value.replace(/\\|'/g, '');
        return props.filterOption(_query, item);
      }
      if (props.checkStrictly) {
        return getFilteredStatus(item.label);
      }
      return item.path?.find((leaf) => getFilteredStatus(leaf.label));
    });
    // if (isFunction(props.filterOption)) {
    //   let _query = '';
    //   _query = computedInputValue.value.replace(/\\|'/g, '');
    //   return props.filterOption(_query, opts);
    // } else {
    //   return opts.filter((item) => {
    //     if (props.checkStrictly) {
    //       return getFilteredStatus(item.label);
    //     }
    //     return item.path?.find((leaf) => getFilteredStatus(leaf.label));
    //   });
    // }
  });
  const updateValue = (values: UnionType[] | UnionType[][]) => {
    const value = props.multiple ? values : values[0] ?? '';
    if (values.length === 0) {
      setSelectedPath();
      setActiveKey();
    }

    _value.value = value;
    emit('update:modelValue', value);
    emit('change', value);
  };

  const handlePopupVisibleChange = (visible: boolean): void => {
    if (computedPopupVisible.value !== visible) {
      menuShow.value = visible;
      emit('popupVisibleChange', visible);
    }
  };

  const handleRemove = (key: string) => {
    if (props.multiple) {
      const option = leafOptionMap.get(key);
      if (option) {
        selectMultiple(option, false);
      } else {
        const values: any[] = [];
        computedValueMap.value.forEach((value, _key) => {
          if (_key !== key) {
            values.push(value);
          }
        });
        updateValue(values);
      }
    }
  };

  const selectSingle = (option: CascaderOptionInfo) => {
    updateValue([props.pathMode ? option.pathValue : option.value]);
    handlePopupVisibleChange(false);
  };

  const selectMultiple = (option: CascaderOptionInfo, checked: boolean) => {
    if (checked) {
      const leafOptionInfos = props.checkStrictly ? [option] : getLeafOptionInfos(option);

      updateValue([
        ...computedValueMap.value.values(),
        ...leafOptionInfos
          .filter((item) => !computedValueMap.value.has(item.key))
          .map((item) => {
            return props.pathMode ? item.pathValue : item.value;
          }),
      ]);
    } else {
      const leafOptionKeys = props.checkStrictly ? [option.key] : getLeafOptionKeys(option);
      const values: any[] = [];
      computedValueMap.value.forEach((value, key) => {
        if (!leafOptionKeys.includes(key)) {
          values.push(value);
        }
      });
      updateValue(values);
    }

    handleInputValueChange('', 'optionChecked');
  };

  const handleClickOption = (option: CascaderOptionInfo, checked?: boolean) => {
    if (props.multiple) {
      selectMultiple(option, checked ?? true);
    } else {
      selectSingle(option);
    }
  };

  const handleSearch = debounce((value: string) => {
    emit('search', value);
  }, props.debounce);

  const handleInputValueChange = (value: string, reason: string): void => {
    if (value !== computedInputValue.value) {
      if (reason === 'manual' && !computedPopupVisible.value) {
        menuShow.value = true;
        emit('popupVisibleChange', true);
      }

      _inputValue.value = value;
      emit('inputValueChange', value);

      if (props.filterable) {
        handleSearch(value);
      }
    }
  };

  watch(computedPopupVisible, (value) => {
    if (value) {
      if (computedValueMap.value.size > 0) {
        const keys = Array.from(computedValueMap.value.keys());
        const lastKey = keys[keys.length - 1];
        const option = leafOptionMap.get(lastKey);
        if (option && option.key !== activeKey.value) {
          setSelectedPath(option.key);
          setActiveKey(option.key);
        }
      }
    } else {
      if (computedValueMap.value.size === 0) {
        setSelectedPath();
        setActiveKey();
      }
      handleInputValueChange('', 'optionListHide');
    }
  });

  const handleClear = (e: MouseEvent) => {
    e.stopPropagation();
    if (props.multiple) {
      // 保留已经被选中但被disabled的选项值
      const newValues: any[] = [];
      computedValueMap.value.forEach((value, key) => {
        const option = leafOptionMap.get(key);
        if (option?.disabled) {
          newValues.push(props.pathMode ? option.pathValue : option.value);
        }
      });
      updateValue(newValues);
    } else {
      updateValue([]);
    }
    handleInputValueChange('', 'manual');
    emit('clear');
  };

  const showSearchPanel = computed(() => props.filterable && computedInputValue.value?.length > 0);

  const handleFocus = (e: FocusEvent) => {
    emit('focus', e);
  };
  const handleBlur = (e: FocusEvent) => {
    emit('blur', e);
  };

  const { activeKey, activeOption, selectedPath, displayColumns, setActiveKey, setSelectedPath, getNextActiveNode } = useSelectedPath(
    optionInfos,
    {
      optionMap,
      filteredLeafOptions,
      showSearchPanel,
      expandChild,
    }
  );

  const handleKeyDown = getKeyDownHandler(
    new Map([
      [
        KEYBOARD_KEY.ENTER,
        (ev: Event) => {
          if (computedPopupVisible.value) {
            if (activeOption.value) {
              let checked: boolean;
              if (props.checkStrictly || activeOption.value.isLeaf) {
                checked = !computedValueMap.value.has(activeOption.value.key);
              } else {
                checked = !getCheckedStatus(activeOption.value, computedValueMap.value).checked;
              }
              setSelectedPath(activeOption.value.key);
              handleClickOption(activeOption.value, checked);
            }
          } else {
            handlePopupVisibleChange(true);
          }
        },
      ],
      [
        KEYBOARD_KEY.ESC,
        (ev: Event) => {
          handlePopupVisibleChange(false);
        },
      ],
      [
        KEYBOARD_KEY.ARROW_DOWN,
        (ev: Event) => {
          ev.preventDefault();
          const activeNode = getNextActiveNode('next');
          setActiveKey(activeNode?.key);
        },
      ],
      [
        KEYBOARD_KEY.ARROW_UP,
        (ev: Event) => {
          ev.preventDefault();
          const activeNode = getNextActiveNode('preview');
          setActiveKey(activeNode?.key);
        },
      ],
      [
        KEYBOARD_KEY.ARROW_RIGHT,
        (ev: Event) => {
          if (!showSearchPanel.value) {
            ev.preventDefault();
            if (activeOption.value?.children) {
              setSelectedPath(activeOption.value.key);
              setActiveKey(activeOption.value.children[0]?.key);
            }
          }
        },
      ],
      [
        KEYBOARD_KEY.ARROW_LEFT,
        (ev: Event) => {
          if (!showSearchPanel.value) {
            ev.preventDefault();
            if (activeOption.value?.parent) {
              setSelectedPath(activeOption.value.parent.key);
              setActiveKey(activeOption.value.parent.key);
            }
          }
        },
      ],
    ])
  );

  const selectViewValue = computed(() => {
    const result: any[] = [];
    computedValueMap.value.forEach((value, key) => {
      const option = leafOptionMap.get(key);
      if (option) {
        result.push({
          value: key,
          option,
          label: props.formatLabel?.(option.path.map((item) => item.raw)) ?? getOptionLabel(option),
          closable: !option.disabled,
          tagProps: option.tagProps,
        });
      }
    });
    return result;
  });

  const showClear = () => {
    if (props.disabled) {
      return;
    }
    showClearable.value = true;
  };
  const hideClear = () => {
    showClearable.value = false;
  };

  onMounted(() => {
    origin.value?.addEventListener('click', openPopup);
  });

  provide(
    cascaderInjectionKey,
    //  @ts-ignore
    reactive({
      onClickOption: handleClickOption,
      setActiveKey,
      setSelectedPath,
      loadMore,
      addLazyLoadOptions,
      formatLabel,
      expandTrigger: trigger,
      slots,
      valueMap: computedValueMap,
    })
  );

  return {
    origin,
    overlayRef,
    menuShow,
    rootClasses,
    menuOpenClass,
    inputValue,
    openPopup,
    rootStyle,
    showClearable,
    position,
    showClear,
    hideClear,
    multiple,

    optionInfos,
    filteredLeafOptions,
    selectedPath,
    activeKey,
    displayColumns,
    computedInputValue,
    computedPopupVisible,
    handleClear,
    selectViewValue,
    handleInputValueChange,
    showSearchPanel,
    handlePopupVisibleChange,
    handleFocus,
    handleBlur,
    handleRemove,
    mergedDisabled,
    totalLevel,
  };
};
