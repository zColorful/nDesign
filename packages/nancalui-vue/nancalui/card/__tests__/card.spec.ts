import { mount } from '@vue/test-utils';
import NCard from '../src/card';
import NAvatar from '../../avatar/src/avatar';
import { useNamespace } from '../../shared/hooks/use-namespace';

const ns = useNamespace('card', true);
const avatar = useNamespace('avatar');

const avatarBaseClass = avatar.b();
const titleClass = ns.e('title');
const subtitleClass = ns.e('subtitle');
const contentClass = ns.e('content');
const actionsClass = ns.e('actions');
const metaClass = ns.e('meta');
const cardBlockClass = '.card-block';
const spaceBetweenClass = `${ns.em('actions', 'align')}-spaceBetween`;
const alwaysShadowClass = ns.em('shadow', 'always');
const hoverShadowClass = ns.em('shadow', 'hover');
const neverShadowClass = ns.em('shadow', 'never');

describe('card', () => {
  it('should render correctly', async () => {
    const wrapper = mount(NCard);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render correctly avatar', async () => {
    const wrapper = mount({
      components: {
        NCard,
        NAvatar,
      },
      template: `
        <n-card class="n-card">
          <template #avatar>
            <n-avatar name="nancalui"></n-avatar>
          </template>
        </n-card>
      `,
    });
    const avatarComponent = wrapper.findAllComponents({ name: 'NAvatar' })[0];
    expect(avatarComponent.classes()).toContain(avatarBaseClass);
  });

  it('in v-slot mode should render correctly avatar', async () => {
    const wrapper = mount({
      components: {
        NCard,
        NAvatar,
      },
      template: `
        <n-card class="n-card">
          <template v-slot:avatar>
            <n-avatar name="nancalui"></n-avatar>
          </template>
        </n-card>
      `,
    });
    const avatarComponent = wrapper.findAllComponents({ name: 'NAvatar' })[0];
    expect(avatarComponent.classes()).toContain(avatarBaseClass);
  });

  it('should render correctly title', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card">
          <template #title>
            nancalui
          </template>
        </n-card>
      `,
    });
    expect(wrapper.find(titleClass).text()).toBe('nancalui');
  });

  it('in v-slot mode should render title', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card">
          <template v-slot:title>
            nancalui Course
          </template>
        </n-card>
      `,
    });
    expect(wrapper.find(titleClass).text()).toBe('nancalui Course');
  });

  it('should render correctly subtitle', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card">
          <template #subtitle>
            nancalui
          </template>
        </n-card>
      `,
    });
    expect(wrapper.find(subtitleClass).text()).toBe('nancalui');
  });

  it('in v-slot mode should render subtitle', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card">
          <template v-slot:subtitle>
            nancalui
          </template>
        </n-card>
      `,
    });
    expect(wrapper.find(subtitleClass).text()).toBe('nancalui');
  });

  it('should render correctly content', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card">
          <template #content>
            nancalui
          </template>
        </n-card>
      `,
    });
    expect(wrapper.find(contentClass).text()).toBe('nancalui');
  });

  it('in v-slot mode should render content', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card">
          <template v-slot:content>
            nancalui
          </template>
        </n-card>
      `,
    });
    expect(wrapper.find(contentClass).text()).toBe('nancalui');
  });

  it('should render correctly actions', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card">
          <template #actions>
            <div class="card-block">
              btn
            </div>
          </template>
        </n-card>
      `,
    });
    expect(wrapper.find(actionsClass).exists()).toBeTruthy();
    expect(wrapper.find(cardBlockClass).text()).toBe('btn');
  });

  it('in v-slot mode should render actions', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card">
          <template #actions>
            <div class="card-block">
              btn
            </div>
          </template>
        </n-card>
      `,
    });
    expect(wrapper.find(actionsClass).exists()).toBeTruthy();
    expect(wrapper.find(cardBlockClass).text()).toBe('btn');
  });

  it('should render correctly image', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card" :src="'https://devui.design/components/assets/image1.png'">
        </n-card>
      `,
    });
    expect(wrapper.find(metaClass).attributes('src').includes('https://devui.design/components/assets/image1.png')).toBeTruthy();
  });

  it('should render correctly align', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card class="n-card" :align="'spaceBetween'">
        </n-card>
      `,
    });
    expect(wrapper.find(spaceBetweenClass).exists()).toBeTruthy();
  });

  it('should render correctly shadow', async () => {
    const wrapper = mount({
      components: {
        NCard,
      },
      template: `
        <n-card >
        </n-card>
      `,
    });
    expect(wrapper.find(hoverShadowClass).exists()).toBeTruthy();

    await wrapper.setProps({
      shadow: 'always',
    });
    expect(wrapper.find(alwaysShadowClass).exists()).toBeTruthy();

    await wrapper.setProps({
      shadow: 'never',
    });
    expect(wrapper.find(neverShadowClass).exists()).toBeTruthy();
  });
});
