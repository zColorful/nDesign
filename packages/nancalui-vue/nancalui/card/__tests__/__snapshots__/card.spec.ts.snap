// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card should render correctly 1`] = `
<div
  class="card-container nancalui-card nancalui-card__shadow--hover"
>
  <!---->
  <div
    class="nancalui-card__header"
  >
    
    <div>
      <div
        class="nancalui-card__title"
      >
        <!---->
      </div>
      <div
        class="nancalui-card__subtitle"
      >
        <!---->
      </div>
    </div>
  </div>
  
  <div
    class="nancalui-card__content"
  >
    <!---->
  </div>
  <div
    class="nancalui-card__actions"
  >
    
  </div>
</div>
`;
